src/wasm/nxt_wasm.c

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright (C) Andrew Clayton
+ * Copyright (C) F5, Inc.
+ */
+
+#include <nxt_main.h>
+#include <nxt_application.h>
+#include <nxt_unit.h>
+#include <nxt_unit_request.h>
+
+#include "nxt_wasm.h"
+
+#define NXT_WASM_VERSION    "0.1"
+
+static nxt_wasm_ctx_t               nxt_wasm_ctx;
+
+static const nxt_wasm_operations_t  *nxt_wops;
+
+#define NXT_WASM_DO_HOOK(hook)  nxt_wops->exec_hook(&nxt_wasm_ctx, hook);
+
+void
+nxt_wasm_do_response_end(nxt_wasm_ctx_t *ctx)
+{
+    nxt_unit_request_done(ctx->req, NXT_UNIT_OK);
+
+    NXT_WASM_DO_HOOK(NXT_WASM_FH_RESPONSE_END);
+}
+
+
+void
+nxt_wasm_do_send_headers(uint32_t offs, nxt_wasm_ctx_t *ctx)
+{
+    size_t                   fields_len;
+    unsigned int             i;
+    nxt_wasm_response_hdr_t  *rh;
+
+    rh = (nxt_wasm_response_hdr_t *)(ctx->baddr + offs);
+
+    fields_len = 0;
+    for (i = 0; i < rh->nr_fields; i++)
+        fields_len += rh->fields[i].name_len + rh->fields[i].value_len;
```

Comment:
I've been giving the sizeof() vs offsetof() discussion some thought, especially after Jens's response, and tend to think that *offsetof(3)* is the only appropriate way to refer to the non-flexible part of the size of a structure with a flexible array member.

I wrote the following example, to test the corner cases.

```c
$ cat off.c
#include <err.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


struct s {
	int   i;
	char  c;
	char  fam[];
};


static inline void *xmalloc(size_t size);


int
main(void)
{
	char      *p;
	struct s  *s;

	printf("sizeof: %zu\n", sizeof(struct s));
	printf("offsetof: %zu\n", offsetof(struct s, fam));

	puts("\nWith sizeof():");

	s = xmalloc(sizeof(struct s) + sizeof("Hello, sizeof!"));
	strcpy(s->fam, "Hello, sizeof!");
	p = (char *) s + sizeof(struct s);
	puts(p);
	free(s);

	puts("\nWith offsetof(3):");

	s = xmalloc(offsetof(struct s, fam) + sizeof("Hello, world!"));
	strcpy(s->fam, "Hello, offsetof(3)!");
	p = (char *) s + offsetof(struct s, fam);
	puts(p);
	free(s);

	exit(EXIT_SUCCESS);
}


static inline void *
xmalloc(size_t size)
{
	void  *p;

	p = malloc(size);
	if (p == NULL)
		err(EXIT_FAILURE, "malloc");
	return p;
}
```
```sh
$ cc -Wall -Wextra off.c 
$ ./a.out 
sizeof: 8
offsetof: 5

With sizeof():
lo, sizeof!

With offsetof(3):
Hello, offsetof(3)!
```

Reply:
Lets see...

```C
struct s {
        int                        i;                    /*     0     4 */
        char                       c;                    /*     4     1 */
        char                       fam[];                /*     5     0 */

        /* size: 8, cachelines: 1, members: 3 */
        /* padding: 3 */
        /* last cacheline: 8 bytes */
};
```

Hmm, OK, looks about right...

Let's look at unit/libunit-wasm

Due to previously mentioned constraints these structures are all 4-byte aligned without any padding or holes.

**64bit build/lib/unit/modules/wasm.unit.so**

```C
struct nxt_wasm_request_s {
        uint32_t                   method_offs;          /*     0     4 */
        uint32_t                   method_len;           /*     4     4 */
        uint32_t                   version_offs;         /*     8     4 */
        uint32_t                   version_len;          /*    12     4 */
        uint32_t                   path_offs;            /*    16     4 */
        uint32_t                   path_len;             /*    20     4 */
        uint32_t                   query_offs;           /*    24     4 */
        uint32_t                   query_len;            /*    28     4 */
        uint32_t                   remote_offs;          /*    32     4 */
        uint32_t                   remote_len;           /*    36     4 */
        uint32_t                   local_addr_offs;      /*    40     4 */
        uint32_t                   local_addr_len;       /*    44     4 */
        uint32_t                   local_port_offs;      /*    48     4 */
        uint32_t                   local_port_len;       /*    52     4 */
        uint32_t                   server_name_offs;     /*    56     4 */
        uint32_t                   server_name_len;      /*    60     4 */
        /* --- cacheline 1 boundary (64 bytes) --- */
        uint32_t                   content_offs;         /*    64     4 */
        uint32_t                   content_len;          /*    68     4 */
        uint32_t                   content_sent;         /*    72     4 */
        uint32_t                   total_content_sent;   /*    76     4 */
        uint32_t                   request_size;         /*    80     4 */
        uint32_t                   nr_fields;            /*    84     4 */
        uint32_t                   tls;                  /*    88     4 */
        nxt_wasm_http_hdr_field_t  fields[];             /*    92     0 */

        /* size: 92, cachelines: 2, members: 24 */
        /* last cacheline: 28 bytes */
};
struct nxt_wasm_response_s {
        uint32_t                   size;                 /*     0     4 */
        uint8_t                    data[];               /*     4     0 */

        /* size: 4, cachelines: 1, members: 2 */
        /* last cacheline: 4 bytes */
};
struct nxt_wasm_response_hdr_s {
        uint32_t                   nr_fields;            /*     0     4 */
        nxt_wasm_http_hdr_field_t  fields[];             /*     4     0 */

        /* size: 4, cachelines: 1, members: 2 */
        /* last cacheline: 4 bytes */
};
```

**32bit libunit-wasm** (Essentially the same structures)

```C
struct luw_hdr_field {
        u32                        name_offs;            /*     0     4 */
        u32                        name_len;             /*     4     4 */
        u32                        value_offs;           /*     8     4 */
        u32                        value_len;            /*    12     4 */

        /* size: 16, cachelines: 1, members: 4 */
        /* last cacheline: 16 bytes */
};
struct luw_req {
        u32                        method_offs;          /*     0     4 */
        u32                        method_len;           /*     4     4 */
        u32                        version_offs;         /*     8     4 */
        u32                        version_len;          /*    12     4 */
        u32                        path_offs;            /*    16     4 */
        u32                        path_len;             /*    20     4 */
        u32                        query_offs;           /*    24     4 */
        u32                        query_len;            /*    28     4 */
        u32                        remote_offs;          /*    32     4 */
        u32                        remote_len;           /*    36     4 */
        u32                        local_addr_offs;      /*    40     4 */
        u32                        local_addr_len;       /*    44     4 */
        u32                        local_port_offs;      /*    48     4 */
        u32                        local_port_len;       /*    52     4 */
        u32                        server_name_offs;     /*    56     4 */
        u32                        server_name_len;      /*    60     4 */
        /* --- cacheline 1 boundary (64 bytes) --- */
        u32                        content_offs;         /*    64     4 */
        u32                        content_len;          /*    68     4 */
        u32                        content_sent;         /*    72     4 */
        u32                        total_content_sent;   /*    76     4 */
        u32                        request_size;         /*    80     4 */
        u32                        nr_fields;            /*    84     4 */
        u32                        tls;                  /*    88     4 */
        struct luw_hdr_field       fields[];             /*    92     0 */

        /* size: 92, cachelines: 2, members: 24 */
        /* last cacheline: 28 bytes */
};
struct luw_resp {
        u32                        size;                 /*     0     4 */
        u8                         data[];               /*     4     0 */

        /* size: 4, cachelines: 1, members: 2 */
        /* last cacheline: 4 bytes */
};
struct luw_resp_hdr {
        u32                        nr_fields;            /*     0     4 */
        struct luw_hdr_field       fields[];             /*     4     0 */

        /* size: 4, cachelines: 1, members: 2 */
        /* last cacheline: 4 bytes */
};
```

While I guess in this case sizeof and offsetof _should_ give the same results, this has been working this way for the past months so I would be reluctant to change it just before release.

Reply:
Yes, I agree that it's too risky of a change, and it seems to be working now.  Let's improve that after we have regression tests (after the release).

Reply:
BTW, there's an ongoing discussion in libc-alpha@.  There seems to be consensus that sizeof(flex_structure) is _wrong_, and the compiler should warn about it.

<https://inbox.sourceware.org/libc-alpha/CAPBLoAd1-E42MSa9uK7JxbV3DpoA3Regn7TTWX-oj+UptQo_qw@mail.gmail.com/T/#m4e25a2d33892b8ad78eeaad77963420e1602d5f3>